airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py

```diff
@@ -172,27 +177,89 @@ def _doc_upload(
     metadata_dict["data"]["dockerImageTag"] = validator_opts.prerelease_tag
     for registry in get(metadata_dict, "data.registries", {}).values():
         if "dockerImageTag" in registry:
             registry["dockerImageTag"] = validator_opts.prerelease_tag
 
-    # write metadata to yaml file in system tmp folder
-    tmp_metadata_file_path = Path("/tmp") / metadata.data.dockerRepository / validator_opts.prerelease_tag / METADATA_FILE_NAME
-    tmp_metadata_file_path.parent.mkdir(parents=True, exist_ok=True)
-    with open(tmp_metadata_file_path, "w") as f:
-        yaml.dump(metadata_dict, f)
+    return metadata_dict
+
+
+def _get_git_info_for_file(original_metadata_file_path: Path) -> Optional[GitInfo]:
+    """
+    Add additional information to the metadata file before uploading it to GCS.
+
+    e.g. The git commit hash, the date of the commit, the author of the commit, etc.
+
+    """
+    repo = git.Repo(search_parent_directories=True)
+    try:
+        # get the commit hash for the last commit that modified the metadata file
+        commit_sha = repo.git.log("-1", "--format=%H", str(original_metadata_file_path))
+        commit_timestamp = repo.git.log("-1", "--format=%ct", str(original_metadata_file_path))
+        commit_author = repo.git.log("-1", "--format=%an", str(original_metadata_file_path))
+        commit_author_email = repo.git.log("-1", "--format=%ae", str(original_metadata_file_path))
+
+        return GitInfo(
+            commit_author=commit_author,
+            commit_author_email=commit_author_email,
+            commit_timestamp=commit_timestamp,
+            commit_sha=commit_sha,
+        )
+    except git.exc.GitCommandError as e:
+        if "unknown revision or path not in the working tree" in str(e):
+            logging.warning(f"Metadata file {original_metadata_file_path} is not tracked by git, skipping author info attachment.")
+            return None
+        else:
+            raise e
+
+
+def _apply_author_info_to_metadata_file(metadata_dict: dict, original_metadata_file_path: Path) -> dict:
+    """Apply author info to the metadata file before uploading it to GCS."""
+    git_info = _get_git_info_for_file(original_metadata_file_path)
+    if git_info:
+        # Apply to the nested / optional field at metadata.data.ab_internal.git
+        git_info_dict = to_json_sanitized_dict(git_info, exclude_none=True)
+        metadata_dict = set_(metadata_dict, "data.ab_internal.git", git_info_dict)
+    return metadata_dict
 
-    return tmp_metadata_file_path
+
+def _write_metadata_to_tmp_file(metadata_dict: dict) -> Path:
+    """Write the metadata to a temporary file."""
+    with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml", delete=False) as tmp_file:
+        yaml.dump(metadata_dict, tmp_file)
+        return Path(tmp_file.name)
+
+
+def _safe_load_metadata_file(metadata_file_path: Path) -> dict:
+    try:
+        metadata = yaml.safe_load(metadata_file_path.read_text())
```

Comment:
Would deserialization to a metadata model be more robust in terms of validation?

Reply:
So the reason I didnt deserialize to a metadata model here is that we techincally dont want to do validation until the metadata file is considered done. i.e. has had all the transforms applied to it. 

This lets us add mandatory generated fields easier in the future.

If thats too future optomized though let me know.